neonvm/tools/vm-builder/main.go

```diff
@@ -250,24 +251,22 @@ fi
 ::respawn:su -p nobody -c '/usr/local/bin/pgbouncer /etc/pgbouncer.ini'
 ::respawn:su -p nobody -c 'DATA_SOURCE_NAME="user=cloud_admin sslmode=disable dbname=postgres" /bin/postgres_exporter --auto-discover-databases --exclude-databases=template0,template1'
 ttyS0::respawn:/neonvm/bin/agetty --8bits --local-line --noissue --noclear --noreset --host console --login-program /neonvm/bin/login --login-pause --autologin root 115200 ttyS0 linux
+::shutdown:/neonvm/bin/vmshutdown
 `
 
 	scriptVmAcpi = `
 event=button/power
-action=/neonvm/bin/powerdown
+action=/neonvm/bin/poweroff
 `
 
-	scriptPowerDown = `#!/neonvm/bin/sh
-
+	scriptVmShutdown = `#!/neonvm/bin/sh
+rm /neonvm/vmstart.allowed
+if [ -e /neonvm/vmstart.allowed ]; then
+	echo "Error: could not remove vmstart.allowed marker, might hang indefinitely during shutdown" >2&
+fi
 su -p postgres --session-command '/usr/local/bin/pg_ctl stop -D /var/db/postgres/compute/pgdata -m fast --wait -t 10'
-# The poweroff below is the busybox poweroff which goes straight to the kernel, i.e., LINUX_REBOOT_CMD_POWER_OFF / RB_POWER_OFF.
-# Our experiments have shown that, generally, this type of hard shutdown will not FIN/RST existing TCP connections (i.e., state ESTABLISHED).
-# Now, for the particular case of NeonVM, we just did 'pg_ctl stop'.
-# But, the libpagestore.c is currently not explicitly shutting down the TCP connection to pageservers, relying on the kernel
-# to implicitly close the socket on exit.
-# This sleep here is to give the kernel time to send the FIN/RST.
-sleep 1
-/neonvm/bin/poweroff
+# wait for ongoing command to exit by grabbing the lock
```

Comment:
I _think_ there's still a race condition here, if I'm not mistaken.

Because `vmstarter.sh` → `postgres` isn't atomic, we can remove `vmstart.allowed` and run `pg_ctl stop` _after_ compute_ctl starts (again), but before postgres does, so pg_ctl does nothing, and we wait on flock until compute_ctl exits on its own, with the signal ignored.

Reply:
You’re correct, will try to model this and come up with a fix.

I think ultimately the only right way to do it is to teach compute_ctl to do proper shutdown on SIGTERM.

Then we’ll send SIGTERM to vmstarter.sh and have it propagate it to the workload.

Reply:
fixed in https://github.com/neondatabase/autoscaling/pull/525/commits/4e58bd38443217f4f093fd1cf19651f3b8846cbc , please resolve this conversation if you're convinced and open a new conversation if the fix introduces new problems / you find other race conditions.